File: prob78.java
// S30 Big N Problem #78 {Medium }
// 146. LRU Cache
// Time Complexity :O(1)
// Space Complexity :O(n) where n is the number of nodes(capacity)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :None


// Your code here along with comments explaining your approach
class LRUCache {
    class Node{
        int key; int value;
        Node prev; Node next;
        public Node(int key,int value){
            this.key=key;
            this.value=value;
        }
    }
    
    private void removeNode(Node node){
        node.next.prev=node.prev;
        node.prev.next=node.next;
    }
    
    private void addToHead(Node node){
        node.prev=head;
        node.next=head.next;
        node.next.prev=node;
        head.next=node;
    }
    
    HashMap<Integer,Node> map;
    Node head; Node tail;
    int capacity;
    public LRUCache(int capacity) {
        this.map=new HashMap<>();
        this.head=new Node(-1,-1);
        this.tail=new Node(-1,-1);
        this.head.next=this.tail;
        this.tail.prev=this.head;
        this.capacity=capacity;
    }
    
    
    
    public int get(int key) {
        if(!map.containsKey(key)) return -1;
        Node node=map.get(key); 
        removeNode(node);
        addToHead(node);
        return node.value;
    }
    
    public void put(int key, int value) {
        //key is already in the cache
        if(map.containsKey(key)){
            Node node=map.get(key);
            node.value=value;
            removeNode(node);
            addToHead(node);
        }else{
            Node newnode=new Node(key,value);
            //if cache is full
            if(map.size()>=capacity){
                Node leastUsed=tail.prev;
                map.remove(leastUsed.key);
                removeNode(leastUsed);
            }
            addToHead(newnode);
            map.put(key,newnode);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: prob_78.py
#LRU Cache
class Node:
    def __init__(self,key=None,val=None):
        self.key_val=(key,val)
        #self.val=val
        self.prev=None
        self.next=None

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.head= Node()
        self.tail=Node()
        self.head.next=self.tail
        self.tail.prev=self.head
        self.head.prev=None
        self.tail.next=None
        self.size=capacity
        self.map={}

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.map:
            node=self.map[key]
            self.deletenode(key)
            self.addnode_left(node)
            return node.key_val[1] # Getting key value if it is present
        return -1



    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.map: #if key already present, then we delete it in the duble linked list and create a new one
            self.deletenode(key)
        elif len(self.map)>=self.size: #if key not present in map but max size of cache reached
            node=self.tail.prev
            self.deletenode(node.key_val[0])
        node=Node(key,value)
        self.addnode_left(node)

    def addnode_left(self,node):
        self.map[node.key_val[0]]=node
        nxt=self.head.next
        self.head.next=node
        node.prev=self.head
        node.next=nxt
        nxt.prev=node



    def deletenode(self,key):
        node=self.map[key]
        prev=node.prev
        next=node.next
        prev.next=next
        next.prev=prev
        del self.map[key] #del in map as well
